### PR TITLE
Put a real big warning on forms that wp isn't defined and form will not work

### LINKED ISSUE #3199: wp is not defined should be caught and show a notice
Show a notice when WP is not defined that explains the known issues

### PR DESCRIPTION
**Pull request by Shelob9**
Fixes #3199

This shows a notice above the form if `window.wp` is not defined, which it may not be - #3141 
Can be cacused by https://fr.wordpress.org/plugins/disable-embeds/ or WP rocket before 3.3.3 https://github.com/wp-media/wp-rocket/issues/1720

I am considering removing the form if this fails, as its not going to work and if `wp` isn't defined, Caldera Forms can't be expected to function properly.

👀 Right now this is the alert ->
Screen Shot 2019-05-16 at 12 05 58 PM

That language is not great. I will think on it and ask WP Buffs folks what they think.

### CODE REVIEW

File: classes/field/localizer.php [hunk 1]
```diff
@@ -104,14 +104,21 @@ public static function localize_cb(){
...
 	protected static function error_strings(){
-		$strings = array(
-			'mixed_protocol' => __( 'Submission URL and current URL protocols do not match. Form may not function properly.', 'caldera-forms' ),
-			'jquery_old'     => __( 'An out of date version of jQuery is loaded on the page. Form may not function properly.', 'caldera-forms' )
-		);
-
-		if( ! current_user_can( Caldera_Forms::get_manage_cap( 'admin' )) ){
-			unset( $strings[ 'jquery_old' ] );
-		}
+        if( current_user_can( Caldera_Forms::get_manage_cap('admin') ) ) {
```
Comment (Shelob9) [inline classes/field/localizer.php]: @New0 This is a bit of a change - less info for people who can't fix the problem, more info for people who can.

File: classes/field/localizer.php [hunk 2]
```diff
@@ -104,14 +104,21 @@ public static function localize_cb(){
...
+                'wp_not_defined' => __( 'The variable wp is not defined, your forms will not work.', 'caldera-forms' ),
+                'mixed_protocol' => __( 'Submission URL and current URL protocols do not match. Form may not function properly.', 'caldera-forms' ),
+                'jquery_old'     => __( 'An out of date version of jQuery is loaded on the page. Form may not function properly.', 'caldera-forms' )
+            );
+        }else{
+            $message =  __( 'Form may not function properly. Contact site administrator.', 'caldera-forms' );
+            $strings = array(
+                'wp_not_defined' => $message,
+                'mixed_protocol' => $message,
+                //jquery_old is intentionally omitted. See assets/js/front-end-script-init.js ~L378.
```
Comment (Shelob9) [inline classes/field/localizer.php]: LOL...

**Comment by Tabrisrp** [comment]
@Shelob9 do you understand why removing the inline emoji script would cause this error in Caldera?

**Comment by Shelob9** [comment]
It causes the `wp` global to be undefined. That is bad as we rely on `wp.element`, which we enqueue using the standard APIs. I don't understand why that happens. I presume that core is pressuming that `wp` is going to be defined when it starts attaching stuff to it in `wp-includes/script-loader.php` Look in `wp_default_scripts()`. How are you removing emojis?

**Comment by Tabrisrp** [comment]
Everything is in there: https://github.com/wp-media/wp-rocket/blob/master/inc/common/emoji.php

When @webtraingwheels tested, she found that it was specifically this line that was causing the issue: https://github.com/wp-media/wp-rocket/blob/master/inc/common/emoji.php#L11

I really fail to see which of these is related to the `wp` global unfortunately.

**Comment by Shelob9** [comment]
@Tabrisrp OK, this makes a lot more sense after seeing what WP Rocket is doing. Also WordPress and/ or Josh are silly.

So, one of the things that disabling emojis is doing is preventing this line of code from running 
* https://github.com/WordPress/WordPress/blob/master/wp-includes/js/wp-emoji.js#L290-L295

That sets up the `wp` variable. We are enqueuing wp-element using a lot of logic on top of normal wp_enqueue/register_script ->
* https://github.com/CalderaWP/Caldera-Forms/blob/master/classes/render/assets.php#L647
* https://github.com/CalderaWP/Caldera-Forms/blob/master/classes/render/assets.php#L665

Are webpack config treats WordPress dependencies as a externals and doesn't bundle them. It assumes that `wp.element` is set. That apparently is not be possible, if core didn't do `window.wp = window.wp || {};`

**Comment by Tabrisrp** [comment]
Ok so what we can do on our side is, if we detect that Caldera forms is active, we prevent the use of the disable emoji option, adding an explanation in our settings page. Looks good for you?

**Comment by Shelob9** [comment]
That would work, I wonder if there is a simpler solution. I also worry about how this effects WP Rocket + other plugins that use WordPress core scripts in front-end. Maybe printing `window.wp = window.wp || {};` in the header where emoji loader would go?

**Comment by Shelob9** [comment]
Also fixes #3132

**Comment by Shelob9** [comment]
Also, fixes #3141

**Pull request merged**